src/silly_sprites/opengl/shader_program.cpp

```diff
@@ -0,0 +1,179 @@
+            switch (type) {
+                case Type::Vertex:
+                    success = compile_single(id, fallback_vertex_source);
+                    if (not success) {
+                        auto message2 = error_message(id);
+                        spdlog::critical(
+                                "ERROR::SHADER::{}::COMPILATION_FAILED\nerror: {}",
+                                get_name_from_type(type),
+                                message2
+                        );
+                        glDeleteShader(id);
+                        throw GLError(
+                                GLErrorType::FailedToCompileVertexShader,
+                                fmt::format("first error: {}\nsecond error: {}", message, message2)
+                        );
+                    }
+                    break;
+
+                case Type::Geometry:
+                    spdlog::critical("ERROR::SHADER::{}::NO_FALLBACK", get_name_from_type(type));
+                    throw GLError(GLErrorType::FailedToCompileGeometryShader, message);
+                    break;
+
+                case Type::Fragment:
+                    success = compile_single(id, fallback_fragment_source);
+                    if (not success) {
+                        auto message2 = error_message(id);
+                        spdlog::critical(
+                                "ERROR::SHADER::{}::COMPILATION_FAILED\nerror: {}",
+                                get_name_from_type(type),
+                                message2
+                        );
+                        glDeleteShader(id);
+                        throw GLError(
+                                GLErrorType::FailedToCompileVertexShader,
+                                fmt::format("first error: {}\nsecond error: {}", message, message2)
+                        );
+                    }
+                    break;
+            }
+        }
+
+        spdlog::info("SUCCESS::SHADER::{}::COMPILATION", get_name_from_type(type));
+        return id;
+    }
+
+    void ShaderProgram::attach_shader(GLuint const shader) const {
+        glAttachShader(m_program_name, shader);
+        glDeleteShader(shader);
+    }
+
+    void ShaderProgram::link_program() const {
+        glLinkProgram(m_program_name);
+
+        GLint success;
+        glGetProgramiv(m_program_name, GL_LINK_STATUS, &success);
+        if (not success) {
+            auto message = std::string{};
+            GLint len;
+            glGetProgramiv(m_program_name, GL_INFO_LOG_LENGTH, &len);
+            message.resize(gsl::narrow<usize>(len - 2));
+            glGetProgramInfoLog(m_program_name, len - 1, nullptr, message.data());
+            spdlog::critical("ERROR::PROGRAMM::LINK_FAILED -> {}\n", message.data());
+            glDeleteProgram(m_program_name);
+            throw GLError(GLErrorType::FailedToLinkShaderProgram, message);
+        }
```

Comment:
This is just like the error message lambda in the other function, so maybe just extract this into a private helper function (maybe `static`?) to avoid duplication. Maybe then we do not even need that function anymore.

Reply:
because of refactoring the "other function" is now in a other class.
should I move it in some kind of helper class or something?
im not quite shure how to heandle it.

Reply:
Maybe create a free function in some of the OpenGL-related files? I think this code is too specific to just put it in the general-purpose utils header. What do you think?

Reply:
and also these are calling different finctions even when they look quite the same.
here `glGetProgramiv()` and `glGetProgramInfoLog()` gets called and in `shader.cpp` the shader functions gets called.
i could move this to `shader_types.hpp` just because both include this. but that sounds kinda wonkey.
or i add a new file for that but that sounds kinda overkill.
and how should i handle the different calls even if the process would be the same. the standart bool like `bool is_programm = false` is always not that nice. is there a other way a can decide what function to call? i could ann an interface and that functions kinda like this.
```
class ShaderProgammDecision {
public:
    [[nodiscard]] virtual bool is_programm() const { return false; }
    [[nodiscard]] virtual bool is_shader() const { return false; }
}
```
then i could aks the 'ShaderProgammDecision' what to call. but i dont like that ether.
maybe there is a way i can check the datatype in this case. something like.
```
switch (entry) {
    case Shader:
        // call shader stuff
    case Programm:
        // call programm stuff
    default:
        // print some error or something
``` 

do you have an idea on that one?

Reply:
Ah okay, I see. I didn't notice that the functions are different. Then maybe leave the code in its current version. I don't think it's worth the time to make this more elegant then.